### PR TITLE
feat(utils): introduce argument value getter

### PR DESCRIPTION
**Pull request by P403n1x87**
## Description

This util helper has been extracted from the original coding of #1148 for reuse in other integrations. This allows for easy access of argument values from wrapped callables, which receive all their args via generic `*args` and `**kwargs`.

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

**Review by jd — CHANGES_REQUESTED** [review]
Should have a test :)

**Comment by P403n1x87** [comment]
> Should have a test :)

yep, working on it :)

**Comment by P403n1x87** [comment]
It goes without saying that, if the signature of the function whose arguments we are trying to infer changes, then we would have problems. One could make the inferrer method slightly smarter by having it peek at `__code__.co_varnames` and `__code__.co_argcount` (and potentially at `__defaults__` too) to handle indices and default values, so that at least one wouldn't have to worry of arguments having different indices across different integration releases.

**Comment by brettlangdon** [comment]
I wonder how far down the path of [PyArg_ParseTupleAndKeywords](https://docs.python.org/3/c-api/arg.html#c.PyArg_ParseTupleAndKeywords) is worth going.

A lot probably depends on the compatibility and failure modes we want to take.

Does the signature have to match exactly in order for us to trace, or do we allow one-off failures (e.g. trace, but omit metadata we didn't end up getting?)

**Comment by P403n1x87** [comment]
> Does the signature have to match exactly in order for us to trace, or do we allow one-off failures (e.g. trace, but omit metadata we didn't end up getting?)

Worth noting that the current implementation of the inferrer hardly ever fails and you always get a value back. This means that if you were trying to get an argument that's total bonkers, you'd still get back a `default` value (which might as well be `None`, giving you the impression that the argument was resolved and its value is actually `None`). So in order to use this inferrer one needs to really know what they're doing.

**Comment by jd** [comment]
> It goes without saying that, if the signature of the function whose arguments we are trying to infer changes, then we would have problems

Agreed, then I'm not really sure how valuable this kind of "magic" is. I lack the initial context (didn't review the PR) but there is no other good solution than supporting all possible signature of a function, past and present. Trying to support future changes (or uncovered past versions) is a recipe for a disaster of missing or wrong data.

Is it better to provide wrong/missing data to a user or to say "hey, don't upgrade yet, we don't support this version"?

(or am I going too far with this? 😆 )

**Comment by brettlangdon** [comment]
This is always an interesting topic. How strict do we adhere to our library support matrix?

We've mostly gone for an approach to support every version we can today easily, but write the code in a way where, as long as the signature doesn't change we will continue to work on new versions... but also, if it does change, don't error out, just do whatever you can because missing metadata is better than no data at all.

However, there is probably an in-between sweet spot we haven't hit yet, like.... do customers know if our tracing doesn't support the newest version or does one day after an upgrade metadata (or maybe even entire traces) just disappear?

**Comment by majorgreys** [comment]
The original use case was for cases where a kwarg is passed as a positional argument assuming a fixed signature. I don't see it as an attempt to support multiple versions and changing function signatures. 

We might want to avoid suggesting we are "inferring" anything since the intent is far more conservative: check kwargs and back off to args. Having a default of `None` here can be confusing since that's used to avoid adding any metadata derived from the argument that's missing. The function should always evaluate to `None` if the argument is not found.

**Comment by Kyle-Verhoog** [comment]
> It goes without saying that, if the signature of the function whose arguments we are trying to infer changes, then we would have problems.

Yup this basically applies to all our monkey-patching.

> Trying to support future changes (or uncovered past versions) is a recipe for a disaster of missing or wrong data.

💯 

> Is it better to provide wrong/missing data to a user or to say "hey, don't upgrade yet, we don't support this version"?

Right now I think we just fail as we assume that by testing with the latest version of libraries that we catch these issues. It's worth thinking about what we should do in this case. Having a helper like this one could immediately tell us if the code we're dealing with is unsupported which we could use to short circuit/log a warning that the version is unsupported.

I'd say it's worth raising a custom exception or returning a sentinel value for when the argument is not found (so we can distinguish between `None` was given vs. nothing was found)

**Comment by jd** [comment]
> I'd say it's worth raising a custom exception or returning a sentinel value for when the argument is not found (so we can distinguish between None was given vs. nothing was found)

👍  for an exception in such cases.

**Comment by P403n1x87** [comment]
> I'd say it's worth raising a custom exception or returning a sentinel value for when the argument is not found (so we can distinguish between `None` was given vs. nothing was found).

The `infer_argument_value` unfortunately has no way of knowing when to raise an exception, unless this is signalled from the outside via, e.g., the `default` argument. For instance, if you do not expect a default and hence want to fail because you do not have either a kwarg nor a posarg, you could pass a `NoDefault` class/object:

~~~ python
# raises custom exception
infer_argument_value([], {"foo": "bar"}, 2, "snafu", NoDefault)
~~~

**Comment by Kyle-Verhoog** [comment]
Yeah if `infer_argument_value` raises when the value is not found then the caller can handle this and set a default value if need-be. Don't think it makes sense to have the function return a default.

`infer_argument_value` should probably be `get_argument_value`

File: ddtrace/utils/__init__.py [hunk 1]
```diff
@@ -25,3 +30,41 @@ def __get__(self, cls, owner):
...
         return classmethod(self.fget).__get__(None, owner)()
+
+
+class ArgumentError(Exception):
+    """
+    This is raised when an argument lookup, either by position or by keyword, is
+    not found.
+    """
+
+    pass
```
Comment (jd) [inline ddtrace/utils/__init__.py:41]:
```suggestion
```

File: ddtrace/utils/__init__.py [hunk 2]
```diff
@@ -25,3 +30,41 @@ def __get__(self, cls, owner):
...
+    :param pos: The positional index of the argument if passed in as a positional arg
+    :param kw: The name of the keyword if passed in as a keyword argument
+    :return: The value of the target argument
+    """
+    try:
+        return kwargs[kw]
+    except KeyError:
+        if pos < len(args):
+            return args[pos]
+    raise ArgumentError("%s (at position %d)" % (kw, pos))
```
Comment (jd) [inline ddtrace/utils/__init__.py:70]:
```suggestion
        try:
            return args[pos]
        except IndexError:
            raise ArgumentError("%s (at position %d)" % (kw, pos))
```

Comment (P403n1x87) [reply ddtrace/utils/__init__.py:70]: I wanted to code it like this in the first place. If @brettlangdon is OK with it I'll make the change 😁

Comment (brettlangdon) [reply ddtrace/utils/__init__.py:68]: Yeah, all good on raising an exception 👍🏻

**Review by Kyle-Verhoog — DISMISSED** [review]
👍 👍

**Review by jd — APPROVED** [review]

**Pull request merged**